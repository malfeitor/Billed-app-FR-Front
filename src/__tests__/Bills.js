/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom'
import {screen, waitFor} from "@testing-library/dom"
import BillsUI from "../views/BillsUI.js"
import { bills } from "../fixtures/bills.js"
import { ROUTES_PATH} from "../constants/routes.js";
import {localStorageMock} from "../__mocks__/localStorage.js";
import mockedBills from '../__mocks__/store.js'
import userEvent from '@testing-library/user-event'

import router from "../app/Router.js";
import Bills from '../containers/Bills.js';
import { formatDate, formatStatus } from "../app/format.js"
import $ from 'jquery'

describe("Given I am connected as an employee", () => {
  describe("When I am on Bills Page", () => {
    test("Then bill icon in vertical layout should be highlighted", async () => {
      Object.defineProperty(window, 'localStorage', { value: localStorageMock })
      window.localStorage.setItem('user', JSON.stringify({
        type: 'Employee'
      }))

      const root = document.createElement("div")
      root.setAttribute("id", "root")
      document.body.append(root)
      router()
      window.onNavigate(ROUTES_PATH.Bills)
      await waitFor(() => screen.getByTestId('icon-window'))
      const windowIcon = screen.getByTestId('icon-window')

      expect(windowIcon).toHaveClass('active-icon')
    })

    test("Then bills should be ordered from earliest to latest", () => {
      document.body.innerHTML = BillsUI({ data: bills })
      const dates = screen.getAllByText(/^(19|20)\d\d[- /.](0[1-9]|1[012])[- /.](0[1-9]|[12][0-9]|3[01])$/i).map(a => a.innerHTML)
      const antiChrono = (a, b) => ((a < b) ? 1 : -1)
      const datesSorted = [...dates].sort(antiChrono)
      expect(dates).toEqual(datesSorted)
    })

    test('Then eye icon should show the bill', () => {
      const onNavigate = (pathname) => {
        document.body.innerHTML = ROUTES({ pathname })
      }
      const billsPage = new Bills({document, onNavigate, mockedBills, localStorageMock})
      const mockHandleClickIconEye = jest.fn(billsPage.handleClickIconEye)
      const modalFile = document.querySelector('#modaleFile')
      $.fn.modal = jest.fn(modalFile.classList.add('show'))

      const iconEyes = screen.getAllByTestId('icon-eye')
      iconEyes.forEach(icon => {
        icon.addEventListener('click', () => mockHandleClickIconEye(icon))
      }) 
      iconEyes.forEach(icon => {
        userEvent.click(icon)
      })

      expect(mockHandleClickIconEye).toHaveBeenCalledTimes(iconEyes.length)
      expect(modalFile).toHaveClass('show')
    })

    test('Then bills should match with fixture', async () => {
      const onNavigate = (pathname) => {
        document.body.innerHTML = ROUTES({ pathname })
      }
      const billsFixture = bills.map(doc => {
        return {
          ...doc,
          date: formatDate(doc.date),
          status: formatStatus(doc.status)
        }
      })
      const billsPage = new Bills({document, onNavigate, mockedBills, localStorageMock})
      billsPage.store = mockedBills
      const billsList = await billsPage.getBills()
      billsFixture.forEach(fixture => {
        expect(billsList).toContainEqual(fixture)
      })
    })
  })
})
