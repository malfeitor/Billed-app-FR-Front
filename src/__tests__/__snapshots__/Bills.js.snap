// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given I am a user connected as Employee When I navigate to the Bills page Then we fetch bills from mock API 1`] = `
Document {
  "location": Location {
    "assign": [Function],
    "hash": "#employee/bills",
    "host": "localhost",
    "hostname": "localhost",
    "href": "http://localhost/#employee/bills",
    "origin": "http://localhost",
    "pathname": "/",
    "port": "",
    "protocol": "http:",
    "reload": [Function],
    "replace": [Function],
    "search": "",
    "toString": [Function],
  },
  Symbol(SameObject caches): Object {
    "childNodes": NodeList [
      DocumentType {},
      <html>
        <head />
        <body
          style="background-color: rgb(14, 90, 229);"
        >
          <div
            id="root"
          >
            
    
            <div
              class="layout"
            >
              
      
        
              <div
                class="vertical-navbar"
                style="height: 120vh;"
              >
                
          
                <div
                  class="layout-title"
                >
                   Billed 
                </div>
                
          
                <div
                  data-testid="icon-window"
                  id="layout-icon1"
                >
                  
            
  
                  <svg
                    height="0.2in"
                    viewBox="0 0 18 18"
                    width="0.2in"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    
    
                    <path
                      d="M 8.00,5.00
            C 8.00,5.00 8.00,1.00 8.00,1.00
              8.00,0.45 7.55,0.00 7.00,0.00
              7.00,0.00 1.00,0.00 1.00,0.00
              0.45,0.00 0.00,0.45 0.00,1.00
              0.00,1.00 0.00,5.00 0.00,5.00
              0.00,5.55 0.45,6.00 1.00,6.00
              1.00,6.00 7.00,6.00 7.00,6.00
              7.55,6.00 8.00,5.55 8.00,5.00 Z
            M 1.00,8.00
            C 1.00,8.00 7.00,8.00 7.00,8.00
              7.55,8.00 8.00,8.45 8.00,9.00
              8.00,9.00 8.00,17.00 8.00,17.00
              8.00,17.55 7.55,18.00 7.00,18.00
              7.00,18.00 1.00,18.00 1.00,18.00
              0.45,18.00 0.00,17.55 0.00,17.00
              0.00,17.00 0.00,9.00 0.00,9.00
              0.00,8.45 0.45,8.00 1.00,8.00 Z
            M 17.00,12.00
            C 17.00,12.00 11.00,12.00 11.00,12.00
              10.45,12.00 10.00,12.45 10.00,13.00
              10.00,13.00 10.00,17.00 10.00,17.00
              10.00,17.55 10.45,18.00 11.00,18.00
              11.00,18.00 17.00,18.00 17.00,18.00
              17.55,18.00 18.00,17.55 18.00,17.00
              18.00,17.00 18.00,13.00 18.00,13.00
              18.00,12.45 17.55,12.00 17.00,12.00 Z
            M 11.00,0.00
            C 11.00,0.00 17.00,0.00 17.00,0.00
              17.55,0.00 18.00,0.45 18.00,1.00
              18.00,1.00 18.00,9.00 18.00,9.00
              18.00,9.55 17.55,10.00 17.00,10.00
              17.00,10.00 11.00,10.00 11.00,10.00
              10.45,10.00 10.00,9.55 10.00,9.00
              10.00,9.00 10.00,1.00 10.00,1.00
              10.00,0.45 10.45,0.00 11.00,0.00 Z"
                      fill="#fff"
                      id="Imported Path"
                      stroke="none"
                      stroke-width="0"
                    />
                    
  
                  </svg>
                  

          
                </div>
                
          
                <div
                  data-testid="icon-mail"
                  id="layout-icon2"
                >
                  
            

                  <svg
                    height="0.177778in"
                    viewBox="0 0 20 16"
                    width="0.222222in"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    
  
                    <path
                      d="M 2.00,0.00
           C 2.00,0.00 18.00,0.00 18.00,0.00
             19.10,0.00 20.00,0.90 20.00,2.00
             20.00,2.00 20.00,14.00 20.00,14.00
             20.00,15.10 19.10,16.00 18.00,16.00
             18.00,16.00 2.00,16.00 2.00,16.00
             0.90,16.00 0.00,15.10 0.00,14.00
             0.00,14.00 0.00,2.00 0.00,2.00
             0.00,0.90 0.90,0.00 2.00,0.00 Z
           M 11.65,11.45
           C 11.65,11.45 18.00,7.00 18.00,7.00
             18.00,7.00 18.00,4.90 18.00,4.90
             18.00,4.90 10.65,10.05 10.65,10.05
             10.26,10.32 9.74,10.32 9.35,10.05
             9.35,10.05 2.00,4.90 2.00,4.90
             2.00,4.90 2.00,7.00 2.00,7.00
             2.00,7.00 8.35,11.45 8.35,11.45
             9.34,12.14 10.66,12.14 11.65,11.45 Z"
                      fill="#fff"
                      id="Imported Path"
                      stroke="none"
                      stroke-width="0"
                    />
                    

                  </svg>
                  

          
                </div>
                
          
                <div
                  id="layout-disconnect"
                >
                  
            

                  <svg
                    height="0.255556in"
                    viewBox="0 0 23 23"
                    width="0.255556in"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    
  
                    <path
                      d="M 12.78,0.00
           C 12.78,0.00 10.22,0.00 10.22,0.00
             10.22,0.00 10.22,12.76 10.22,12.76
             10.22,12.76 12.78,12.76 12.78,12.76
             12.78,12.76 12.78,0.00 12.78,0.00 Z
           M 18.95,2.77
           C 18.95,2.77 17.14,4.58 17.14,4.58
             18.17,5.41 19.00,6.46 19.58,7.66
             20.15,8.85 20.44,10.16 20.44,11.48
             20.44,13.85 19.50,16.13 17.83,17.80
             16.15,19.48 13.87,20.42 11.50,20.42
             9.66,20.42 7.86,19.85 6.35,18.79
             4.84,17.72 3.70,16.22 3.08,14.48
             2.46,12.75 2.39,10.86 2.89,9.09
             3.39,7.31 4.42,5.73 5.86,4.57
             5.86,4.57 4.05,2.77 4.05,2.77
             2.90,3.75 1.96,4.94 1.27,6.28
             0.58,7.62 0.16,9.09 0.04,10.59
             -0.08,12.10 0.09,13.61 0.55,15.05
             1.02,16.48 1.76,17.81 2.74,18.96
             6.85,23.79 14.11,24.37 18.95,20.26
             20.22,19.18 21.25,17.84 21.94,16.32
             22.64,14.80 23.00,13.15 23.00,11.48
             23.00,8.13 21.52,4.94 18.95,2.77 Z"
                      fill="#fff"
                      id="Imported Path"
                      stroke="none"
                      stroke-width="0"
                    />
                    

                  </svg>
                  

          
                </div>
                
      
              </div>
              
        
      
              <div
                class="content"
              >
                
        
                <div
                  class="content-header"
                >
                  
          
                  <div
                    class="content-title"
                  >
                     Mes notes de frais 
                  </div>
                  
          
                  <button
                    class="btn btn-primary"
                    data-testid="btn-new-bill"
                    type="button"
                  >
                    Nouvelle note de frais
                  </button>
                  
        
                </div>
                
        
                <div
                  id="data-table"
                >
                  
        
                  <table
                    class="table table-striped"
                    id="example"
                    style="width:100%"
                  >
                    
          
                    <thead>
                      
              
                      <tr>
                        
                
                        <th>
                          Type
                        </th>
                        
                
                        <th>
                          Nom
                        </th>
                        
                
                        <th>
                          Date
                        </th>
                        
                
                        <th>
                          Montant
                        </th>
                        
                
                        <th>
                          Statut
                        </th>
                        
                
                        <th>
                          Actions
                        </th>
                        
              
                      </tr>
                      
          
                    </thead>
                    
          
                    <tbody
                      data-testid="tbody"
                    >
                      
            
    
                      <tr>
                        
      
                        <td>
                          Restaurants et bars
                        </td>
                        
      
                        <td>
                          test2
                        </td>
                        
      
                        <td>
                          2 Fév. 02
                        </td>
                        
      
                        <td>
                          200 €
                        </td>
                        
      
                        <td>
                          Refused
                        </td>
                        
      
                        <td>
                          
        
                          <div
                            class="icon-actions"
                          >
                            
      
                            <div
                              data-bill-url="https://test.storage.tld/v0/b/billable-677b6.a…f-1.jpg?alt=media&token=4df6ed2c-12c8-42a2-b013-346c1346f732"
                              data-testid="icon-eye"
                              id="eye"
                            >
                              
      
                              <svg
                                height="0.166667in"
                                viewBox="0 0 22 15"
                                width="0.244444in"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                

                                <path
                                  d="M 11.00,0.00
      C 6.00,0.00 1.73,3.11 0.00,7.50
        1.73,11.89 6.00,15.00 11.00,15.00
        16.00,15.00 20.27,11.89 22.00,7.50
        20.27,3.11 16.00,0.00 11.00,0.00 Z
      M 11.00,12.50
      C 8.24,12.50 6.00,10.26 6.00,7.50
        6.00,4.74 8.24,2.50 11.00,2.50
        13.76,2.50 16.00,4.74 16.00,7.50
        16.00,10.26 13.76,12.50 11.00,12.50 Z
      M 11.00,4.50
      C 9.34,4.50 8.00,5.84 8.00,7.50
        8.00,9.16 9.34,10.50 11.00,10.50
        12.66,10.50 14.00,9.16 14.00,7.50
        14.00,5.84 12.66,4.50 11.00,4.50 Z"
                                  fill="#0D5AE5"
                                  id="Imported Path"
                                  stroke="none"
                                  stroke-width="0"
                                />
                                

                              </svg>
                              
      
                            </div>
                            
    
                          </div>
                          
      
                        </td>
                        
    
                      </tr>
                      
    
    
                      <tr>
                        
      
                        <td>
                          Services en ligne
                        </td>
                        
      
                        <td>
                          test3
                        </td>
                        
      
                        <td>
                          3 Mar. 03
                        </td>
                        
      
                        <td>
                          300 €
                        </td>
                        
      
                        <td>
                          Accepté
                        </td>
                        
      
                        <td>
                          
        
                          <div
                            class="icon-actions"
                          >
                            
      
                            <div
                              data-bill-url="https://test.storage.tld/v0/b/billable-677b6.a…dur.png?alt=media&token=571d34cb-9c8f-430a-af52-66221cae1da3"
                              data-testid="icon-eye"
                              id="eye"
                            >
                              
      
                              <svg
                                height="0.166667in"
                                viewBox="0 0 22 15"
                                width="0.244444in"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                

                                <path
                                  d="M 11.00,0.00
      C 6.00,0.00 1.73,3.11 0.00,7.50
        1.73,11.89 6.00,15.00 11.00,15.00
        16.00,15.00 20.27,11.89 22.00,7.50
        20.27,3.11 16.00,0.00 11.00,0.00 Z
      M 11.00,12.50
      C 8.24,12.50 6.00,10.26 6.00,7.50
        6.00,4.74 8.24,2.50 11.00,2.50
        13.76,2.50 16.00,4.74 16.00,7.50
        16.00,10.26 13.76,12.50 11.00,12.50 Z
      M 11.00,4.50
      C 9.34,4.50 8.00,5.84 8.00,7.50
        8.00,9.16 9.34,10.50 11.00,10.50
        12.66,10.50 14.00,9.16 14.00,7.50
        14.00,5.84 12.66,4.50 11.00,4.50 Z"
                                  fill="#0D5AE5"
                                  id="Imported Path"
                                  stroke="none"
                                  stroke-width="0"
                                />
                                

                              </svg>
                              
      
                            </div>
                            
    
                          </div>
                          
      
                        </td>
                        
    
                      </tr>
                      
    
    
                      <tr>
                        
      
                        <td>
                          Transports
                        </td>
                        
      
                        <td>
                          test1
                        </td>
                        
      
                        <td>
                          1 Jan. 01
                        </td>
                        
      
                        <td>
                          100 €
                        </td>
                        
      
                        <td>
                          Refused
                        </td>
                        
      
                        <td>
                          
        
                          <div
                            class="icon-actions"
                          >
                            
      
                            <div
                              data-bill-url="https://test.storage.tld/v0/b/billable-677b6.a…61.jpeg?alt=media&token=7685cd61-c112-42bc-9929-8a799bb82d8b"
                              data-testid="icon-eye"
                              id="eye"
                            >
                              
      
                              <svg
                                height="0.166667in"
                                viewBox="0 0 22 15"
                                width="0.244444in"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                

                                <path
                                  d="M 11.00,0.00
      C 6.00,0.00 1.73,3.11 0.00,7.50
        1.73,11.89 6.00,15.00 11.00,15.00
        16.00,15.00 20.27,11.89 22.00,7.50
        20.27,3.11 16.00,0.00 11.00,0.00 Z
      M 11.00,12.50
      C 8.24,12.50 6.00,10.26 6.00,7.50
        6.00,4.74 8.24,2.50 11.00,2.50
        13.76,2.50 16.00,4.74 16.00,7.50
        16.00,10.26 13.76,12.50 11.00,12.50 Z
      M 11.00,4.50
      C 9.34,4.50 8.00,5.84 8.00,7.50
        8.00,9.16 9.34,10.50 11.00,10.50
        12.66,10.50 14.00,9.16 14.00,7.50
        14.00,5.84 12.66,4.50 11.00,4.50 Z"
                                  fill="#0D5AE5"
                                  id="Imported Path"
                                  stroke="none"
                                  stroke-width="0"
                                />
                                

                              </svg>
                              
      
                            </div>
                            
    
                          </div>
                          
      
                        </td>
                        
    
                      </tr>
                      
    
    
                      <tr>
                        
      
                        <td>
                          Hôtel et logement
                        </td>
                        
      
                        <td>
                          encore
                        </td>
                        
      
                        <td>
                          4 Avr. 04
                        </td>
                        
      
                        <td>
                          400 €
                        </td>
                        
      
                        <td>
                          En attente
                        </td>
                        
      
                        <td>
                          
        
                          <div
                            class="icon-actions"
                          >
                            
      
                            <div
                              data-bill-url="https://test.storage.tld/v0/b/billable-677b6.a…f-1.jpg?alt=media&token=c1640e12-a24b-4b11-ae52-529112e9602a"
                              data-testid="icon-eye"
                              id="eye"
                            >
                              
      
                              <svg
                                height="0.166667in"
                                viewBox="0 0 22 15"
                                width="0.244444in"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                

                                <path
                                  d="M 11.00,0.00
      C 6.00,0.00 1.73,3.11 0.00,7.50
        1.73,11.89 6.00,15.00 11.00,15.00
        16.00,15.00 20.27,11.89 22.00,7.50
        20.27,3.11 16.00,0.00 11.00,0.00 Z
      M 11.00,12.50
      C 8.24,12.50 6.00,10.26 6.00,7.50
        6.00,4.74 8.24,2.50 11.00,2.50
        13.76,2.50 16.00,4.74 16.00,7.50
        16.00,10.26 13.76,12.50 11.00,12.50 Z
      M 11.00,4.50
      C 9.34,4.50 8.00,5.84 8.00,7.50
        8.00,9.16 9.34,10.50 11.00,10.50
        12.66,10.50 14.00,9.16 14.00,7.50
        14.00,5.84 12.66,4.50 11.00,4.50 Z"
                                  fill="#0D5AE5"
                                  id="Imported Path"
                                  stroke="none"
                                  stroke-width="0"
                                />
                                

                              </svg>
                              
      
                            </div>
                            
    
                          </div>
                          
      
                        </td>
                        
    
                      </tr>
                      
    
          
                    </tbody>
                    
          
                  </table>
                  
        
                </div>
                
      
              </div>
              
      
    
              <div
                aria-hidden="true"
                aria-labelledby="exampleModalCenterTitle"
                class="modal fade"
                id="modaleFile"
                role="dialog"
                tabindex="-1"
              >
                
      
                <div
                  class="modal-dialog modal-dialog-centered modal-lg"
                  role="document"
                >
                  
        
                  <div
                    class="modal-content"
                  >
                    
          
                    <div
                      class="modal-header"
                    >
                      
            
                      <h5
                        class="modal-title"
                        id="exampleModalLongTitle"
                      >
                        Justificatif
                      </h5>
                      
            
                      <button
                        aria-label="Close"
                        class="close"
                        data-dismiss="modal"
                        type="button"
                      >
                        
              
                        <span
                          aria-hidden="true"
                        >
                          ×
                        </span>
                        
            
                      </button>
                      
          
                    </div>
                    
          
                    <div
                      class="modal-body"
                    >
                      
          
                    </div>
                    
        
                  </div>
                  
      
                </div>
                
    
              </div>
              
  
    
            </div>
          </div>
        </body>
      </html>,
    ],
    "implementation": DOMImplementation {},
  },
}
`;
